javascript/ql/src/semmle/javascript/CharacterEscapes.qll

```diff
@@ -0,0 +1,88 @@
+/**
+ * Provides classes for reasoning about character escapes in literals.
+ */
+
+import javascript
+
+module CharacterEscapes {
+  /**
+   * Provides sets of characters (as strings) with specific string/regexp characteristics.
+   */
+  private module Sets {
+    string ordinaryEscapeChars() { result = "fnrtvux0\\" }
+
+    string regexpCharRangeChars() { result = "bBcdDsSwW" }
+
+    string regexpBackreferenceChars() { result = "123456789k" }
+
+    string regexpMetaChars() { result = "^$*+?.()|{}[]-" }
+  }
+
+  /**
+   * Gets the `i'th` character of `raw`, which is preceeded by an uneven number of backslashes.
+   */
+  bindingset[raw]
+  string getAnEscapedCharacter(string raw, int i) {
+    result = raw.regexpFind("(?<=(^|[^\\\\])\\\\(\\\\{2}){0,10}).", _, i)
```

Comment:
The restriction to 10 doesn't matter in practice, but I think you can avoid it by looking for maximal backslash sequences of odd length and then taking the next character:

```
exists(string s |
  s = raw.regexpFind("(?<!\\\\)\\\\(\\\\{2})*(?!\\\\)", _, i-s.length()) and
  result = raw.charAt(i)
)
```

Reply:
I have tried several formulations of this regular expression. The current lookbehind formulation is the only one that scales. 
I tested your formulation, and terminated it after a few minutes. The current formulation takes approximately a second on the same benchmark.

Reply:
Sure, the current formulation is fine.